### ISSUE TITLE
Value-Parameterized Tests support?

### ISSUE THREAD

**Issue by fire** [bitwes/Gut#114]
Does Gut support Value-Parameterized Tests?

https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests

**Comment by bitwes** [comment]
It does not but this would be a nice thing to have.  I've not used this technique before but I've heard of it.  I'll keep this around for a future enhancement

**Comment by fire** [comment]
I have a use case where I have to one way is to generate test case and the other way is to use Value-Parameterized Tests.

So I don't want to code generate the test cases :/

**Comment by fire** [comment]
The problem is I want to generate test cases for all the tscns in the project and connect a signal from the tscn to the test runner and I'm not sure how.

**Comment by bitwes** [comment]
What are the tests you want to generate? There might be a less elegant way to do this with the tool as it is now.

**Comment by fire** [comment]
A brief outline of the inefficient version:

* Given a random collection of tscns in a project
* Generate a test case for each tscn in a specific folder and its subfolders
* In the test case connect the runner to a signal in the tscn
* run the tests to respond to the signal (We assume every test will emit a signal in a finite amount of time)

**Comment by bitwes** [comment]
Ok.  I've got some ideas on how this might work.  I tried typing code on my phone and it is too annoying so I'll psot back later when I have a chance to sit down at a computer.

**Comment by bitwes** [comment]
This is off the top of my head, not tested, and is specific to what you've described.  It also doesn't help with the more generic feature of  Value-Parameterized Tests.  That being said, this is a quick stab at how I would handle this issue.
```
var   scene_paths = [] # fill this somehow in the before_all or before_each

func before_all():
  # fill scene_paths with a hardcoded list or by traversing the directory where
  # all the scenes are or whatever.

func test_scene_signals():
  for i in range(scene_paths.size()):
    var inst = load(scene_paths[i]).instance()
    watch_signals(inst)

    # do something here with your instance

    # This line will yield until inst emits the signal or 5 seconds has 
    # elapsed.  You'll probably want to change the time.  yield_to is a 
    # utility method in test.gd and YIELD is a constant in test.gd
    yield(yield_to(inst, 'signal_name', 5), YIELD)

    assert_signal_emitted(inst, 'signal_name')
```
If you haven't read these already, these links will be helpful.
https://github.com/bitwes/Gut/wiki/Yielding
https://github.com/bitwes/Gut/wiki/Methods#-watch_signalsobject
Keep reading after `watch_signals` for more methods related to making asserts based on signals.

**Comment by fire** [comment]
I will try the code later today.

**Comment by Xrayez** [comment]
Oh so I was completely unaware of this technique and I've been doing something similar with GUT already:

```gdscript
# test_foobar.gd

extends "res://addons/gut/test.gd"

func test_foobar():
	var scene_path = get_script().get_path().get_base_dir().plus_file('test_foobar.tscn')
	var test_instance = load(scene_path).instance()
	add_child(test_instance)

	yield(yield_to(test_instance, 'success', 1.5), YIELD)
	
	assert_true(test_instance.value == test_instance.value_control)
```

The test script's path could determine the path of the test scene. Forgot about actually being able to assert whether a signal is emitted, this should simplify it.

**Comment by Xrayez** [comment]
Ok, so I'm expanding upon the ideas here, got the following to work locally. `test_foobar.gd` will test all scenes within the same folder if `SCENE_BASEPATH` is empty.

The test scenes should be created in a way to:
* emit `passed` signal if everything goes smooth.
* optionally, emit `failed` signal if something terrible happens (even if all scene's  inner tests passed).
* the time it takes for the scene to verify that everything works (by providing `get_time_limit()` method).

```gdscript
# test_foobar.gd
extends "res://addons/gut/test.gd"

const PASSED_SIGNAL = 'passed'
const FAILED_SIGNAL = 'failed'

const GET_TIME_LIMIT = 'get_time_limit'

const SCENE_PREFIX = 'test_'
const SCENE_BASEPATH = String()
const SCENE_EXTENSIONS = ['tscn', 'scn']

var scenes = []


func before_all():
	scenes = fetch_scenes()


func fetch_scenes():
	var base_path
	if not SCENE_BASEPATH.empty():
		base_path = SCENE_BASEPATH
	else:
		base_path = get_script().get_path().get_base_dir()
	
	var dir = Directory.new()
	dir.open(base_path)
	dir.list_dir_begin(true)

	var item = dir.get_next()

	var scenes_paths = []

	while not item.empty():
		if item.begins_with(SCENE_PREFIX) and not dir.current_is_dir():
			if item.get_extension() in SCENE_EXTENSIONS:
				scenes_paths.append(base_path.plus_file(item))
		item = dir.get_next()

	dir.list_dir_end()
	
	var scenes = []
	for path in scenes_paths:
		var scene = load(path)
		scenes.append(scene)
	
	return scenes
	
	
func test_foobar():
	for scene in scenes:
		var test = scene.instance()

		add_child(test)
		watch_signals(test)

		gut.p("Testing: " + test.name)
		
		var seconds = 1.0 
		if test.has_method(GET_TIME_LIMIT):
			seconds = test.call(GET_TIME_LIMIT)

		yield(yield_for(seconds), YIELD)

		assert_signal_emitted(test, PASSED_SIGNAL)
		assert_signal_not_emitted(test, FAILED_SIGNAL)

		test.queue_free()
```
This is how it looks like when run:
![test_scenes](https://user-images.githubusercontent.com/17108460/59103804-812e9b80-8938-11e9-8405-a2f2bb91fd40.png)

The downside of this is having no clue what actually caused a scene to fail, but it should be easy to figure out why as you can simply run the test scene manually without GUT and see for yourself, so this is really more suitable for speeding up tests creation based on scenes+code rather than pure code, and encourages to write test scenes for new functionality as you start with a scene anyway (Test-driven development?).

**Comment by bitwes** [comment]
Implementation idea.  Allow tests to take parameters and come up with a naming convention for a function that would supply those parameters.  First thought is that it would have the same name as the test but be prefixed with `parameters` instead of `test`.  The method metadata has the parameters in it, so GUT could detect a test with parameters and look for a parameter method.  If it finds one then loop through the values returned from the parameter method calling the test once with each value.

It would look something like:
```
func parameters_foo():
  return [1, 3, 5, 7, 9]

func test_foo(value):
  pass 

func parameters_bar():
  return [['one', 'two'], ['three', 'four']]

func test_bar(val1, val2):
  pass
```
GUT would have to detect and error or fail if a parameter method couldn't be found or if the parameters didn't match.

Step two would be adding helper methods for generating values.

**Comment by bitwes** [comment]
Still thinking about this and I still don't know enough about how/when to use it.  I will note that the above might be an ok approach for a single test that should have parameters but would be tedious for a set of tests that take parameters which I think is something this feature would have to do.

**Comment by bitwes** [comment]
New implementation idea:

Syntax
```
var foo_params = [[1, 2, 3], ['a', 'b', 'c']]
func test_foo(params=use_parameters('foo_params')):
  var foo = Foo.new()
  var result = foo.do_something(params[0], params[1])
  assert_eq(result, params[2])  

func generate_bar_params():
  # some logic to create parameter array
  return the_array_we_made_above

func test_bar(params=use_parameters('generate_bar_params')):
  # make assertions using params like in the other test
  pass
```

GUT can detect when a method has parameters (it cannot get the names of those parameters or default values).  When it detects parameters it will start a loop to call the test multiple times.  On the first call to `use_parameters` it will splice up the parameters and set a counter for how many times GUT should loop, then it will return the first set of parameters.  Each subsequent call will return the next parameters.  GUT will have to detect if `use_parameters` was called after the first iteration of the loop.  That's about the only way to enforce the syntax of using `use_parameters` and avoiding tests that shouldn't have parameters.

There is a lot of room for helper methods and data structures for the parameters so that things are more readable.

**Comment by bitwes** [comment]
Implemented in 7.0.0.  Check out the Parameterized Tests wiki page for more information.
